File: leetcode_53/solution.py
class Solution:
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        _sum, _max = 0, nums[0]
        for x in nums:
            _sum += x
            _sum = max(_sum, x)
            if _sum > _max:
                _max = _sum
        return _max

File: leetcode_53/solution.go
package main

func maxSubArray(nums []int) int {
	n := len(nums)
	if n == 0 {
		return 0
	}
	sum, max := 0, nums[0]
	for _, x := range nums {
		sum += x
		if x > sum {
			sum = x
		}
		if sum > max {
			max = sum
		}
	}
	return max
}

//func maxSubArray(nums []int) int {
//	if len(nums) == 0 {
//		return 0
//	}
//
//	dp := make([]int, len(nums))
//	dp[0] = nums[0]
//
//	max := nums[0]
//	for i := 1; i < len(nums); i++ {
//		dp[i] = dp[i - 1] + nums[i]
//		if dp[i] < nums[i] {
//			dp[i] = nums[i]
//		}
//		if dp[i] > max {
//			max = dp[i]
//		}
//	}
//
//	return max
//}

func main()  {

}